crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+                    visit_chunk_sub_range(
+                        &chunk,
+                        item.component_name,
+                        timeline,
+                        pointer_time_range,
+                        |_, _, num_events| {
+                            hovered_events += num_events;
+                        },
+                    );
+                    num_hovered_messages = hovered_events;
+                    hovered_time_range = hovered_time_range.union(pointer_time_range);
+                }
+            }
+        }
+    };
+
+    // Collect all relevant chunks in the visible time range.
+    // We do this as a separate step so that we can also deduplicate chunks.
+    let visible_time_range = time_ranges_ui
+        .time_range_from_x_range((row_rect.left() - MARGIN_X)..=(row_rect.right() + MARGIN_X));
+    let mut chunk_ranges: Vec<(Arc<Chunk>, ResolvedTimeRange, usize)> = vec![];
+
+    visit_relevant_chunks(
+        db,
+        &item.entity_path,
+        item.component_name,
+        timeline,
+        visible_time_range,
+        |chunk, chunk_timeline, num_events| {
+            chunk_ranges.push((chunk, chunk_timeline.time_range(), num_events));
+        },
+    );
+
+    for (chunk, time_range, num_events) in chunk_ranges {
+        add_data_point(chunk, time_range, num_events);
+    }
+
+    let hovered_x_range = (time_ranges_ui
+        .x_from_time_f32(hovered_time_range.min().into())
+        .unwrap_or(f32::MAX)
+        - MARGIN_X)
+        ..=(time_ranges_ui
+            .x_from_time_f32(hovered_time_range.max().into())
+            .unwrap_or(f32::MIN)
+            + MARGIN_X);
+
+    density_graph.buckets = smooth(&density_graph.buckets);
+
+    density_graph.paint(
+        data_density_graph_painter,
+        row_rect.y_range(),
+        time_area_painter,
+        graph_color(ctx, &item.to_item(), ui),
+        hovered_x_range.clone(),
+    );
+
+    if num_hovered_messages > 0 {
+        ctx.selection_state().set_hovered(item.to_item());
+
+        if time_area_response.clicked_by(egui::PointerButton::Primary) {
+            ctx.selection_state().set_selection(item.to_item());
+            time_ctrl.set_time(hovered_time_range.min());
+            time_ctrl.pause();
+        } else if ui.ctx().dragged_id().is_none() {
+            egui::show_tooltip_at_pointer(
+                ui.ctx(),
+                ui.layer_id(),
+                egui::Id::new("data_tooltip"),
+                |ui| {
+                    show_row_ids_tooltip(
+                        ctx,
+                        ui,
+                        time_ctrl,
+                        db,
+                        item,
+                        hovered_time_range,
+                        num_hovered_messages,
+                    );
+                },
+            );
+        }
+    }
+}
+
+// if sorted:
+// - binary search, show hovered things
+// - if its small enough, full scan
+// - otherwise, hover full chunk
```

Comment:
Is that a TODO or something?

Reply:
Leftover from discussion with Emil, resulting in https://github.com/rerun-io/rerun/pull/6847#discussion_r1673948501